File: Semester4/Systems for Design and Implementation/mpp-proiect-csharp-raulfrandes/persistence/dbRepositories/ParticipantDbRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using log4net;
using model;
using model.validator;
using persistence.interfaces;

namespace persistence.dbRepositories;

public class ParticipantDbRepository: IParticipantRepository
{
    private static readonly ILog Log = LogManager.GetLogger("ParticipantDBRepository");

    private IDictionary<String, string> _props;
    private IValidator<Participant> _participantValidator;

    public ParticipantDbRepository(IDictionary<string, string> props, IValidator<Participant> participantValidator)
    {
        Log.Info("Creating ParticipantDbRepository");
        this._props = props;
        this._participantValidator = participantValidator;
    }
    
    public IEnumerable<Participant> FindAll()
    {
        Log.InfoFormat("Entering FindAll");
        IDbConnection connection = DbUtils.GetConnection(_props);
        IList<Participant> participants = new List<Participant>();
        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from participants";

                using (var dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        long id = dataReader.GetInt64(0);
                        string name = dataReader.GetString(1);
                        string code = dataReader.GetString(2);
                        Participant participant = new Participant(name, code);
                        participant.Id = id;
                        participants.Add(participant);
                    }
                }
            }

            Log.InfoFormat("Exiting FindAll with value {0}", participants);
            return participants;
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    public Participant? FindById(long entityId)
    {
        Log.InfoFormat("Entering FindById with value {0}", entityId);
        IDbConnection connection = DbUtils.GetConnection(_props);

        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from participants where id = @id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = entityId;
                command.Parameters.Add(paramId);

                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        long id = dataReader.GetInt64(0);
                        string name = dataReader.GetString(1);
                        string code = dataReader.GetString(2);
                        Participant participant = new Participant(name, code);
                        participant.Id = id;
                        Log.InfoFormat("Exiting FindById with value {0}", participant);
                        return participant;
                    }
                    Log.InfoFormat("Exiting FindById with value {0}", null);
                    return null;
                }
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    public Participant? Save(Participant entity)
    {
        _participantValidator.Validate(entity);
        Log.InfoFormat("Entering Save with value {0}", entity);
        IDbConnection connection = DbUtils.GetConnection(_props);
        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText =
                    "insert into participants (name, code) values (@name, @code)";

                IDbDataParameter paramName = command.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = entity.Name;
                command.Parameters.Add(paramName);

                IDbDataParameter paramCode = command.CreateParameter();
                paramCode.ParameterName = "@code";
                paramCode.Value = entity.Code;
                command.Parameters.Add(paramCode);

                var rows = command.ExecuteNonQuery();
                Log.InfoFormat("saved {0} instances", rows);

                if (rows > 0)
                {
                    command.CommandText = "select last_insert_rowid()";
                    var result = command.ExecuteScalar();
                    if (result != null && result != DBNull.Value)
                    {
                        long id = Convert.ToInt64(result);
                        entity.Id = id;
                        Log.InfoFormat("Exiting Save with value {0}", entity);
                        return entity;
                    }
                }
            }
            Log.InfoFormat("Exiting Save with value {0}", null);
            return null;
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    public Participant DeleteById(long entityId)
    {
        throw new NotImplementedException();
    }

    public Participant Update(Participant entity)
    {
        throw new NotImplementedException();
    }
}

File: Semester4/Systems for Design and Implementation/mpp-proiect-java-raulfrandes/TriathlonPersistence/src/main/java/triathlon/persistence/dbRepositories/ParticipantDBRepository.java
package triathlon.persistence.dbRepositories;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import triathlon.model.Participant;
import triathlon.model.validator.Validator;
import triathlon.persistence.interfaces.IParticipantRepository;
import triathlon.persistence.RepositoryException;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ParticipantDBRepository implements IParticipantRepository {

    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();
    private final Validator<Participant> participantValidator;

    public ParticipantDBRepository(Properties props, Validator<Participant> participantValidator) {
        logger.info("Initializing ParticipantDBRepository with properties: {}", props);
        this.dbUtils = new JdbcUtils(props);
        this.participantValidator = participantValidator;
    }

    @Override
    public Iterable<Participant> findAll() {
        logger.traceEntry();
        Connection connection = dbUtils.getConnection();
        List<Participant> participants = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection
                .prepareStatement("select * from participants")){
            ResultSet resultSet = preparedStatement.executeQuery();
            while (resultSet.next()) {
                Long id = resultSet.getLong("id");
                String name = resultSet.getString("name");
                String code = resultSet.getString("code");
                Participant participant = new Participant(name, code);
                participant.setId(id);
                participants.add(participant);
            }
            logger.traceExit(participants);
            return participants;
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    @Override
    public Participant findById(Long entityId) {
        if (entityId == null) {
            throw new RepositoryException("Id cannot be null");
        }
        logger.traceEntry("searching participant with id {}", entityId);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection
                .prepareStatement("select * from participants where id = ?")) {
            preparedStatement.setLong(1, entityId);
            ResultSet resultSet = preparedStatement.executeQuery();
            if (resultSet.next()) {
                Long id = resultSet.getLong("id");
                String name = resultSet.getString("name");
                String code = resultSet.getString("code");
                Participant participant = new Participant(name, code);
                participant.setId(id);
                logger.traceExit(participant);
                return participant;
            }
            logger.traceExit(null);
            return null;
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    @Override
    public Participant save(Participant entity) {
        if (entity == null) {
            throw new RepositoryException("Entity cannot be null");
        }
        participantValidator.validate(entity);
        logger.traceEntry("saving participant {}", entity);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection
                .prepareStatement("insert into participants (name, code) values (?, ?)",
                        Statement.RETURN_GENERATED_KEYS)) {
            preparedStatement.setString(1, entity.getName());
            preparedStatement.setString(2, entity.getCode());
            int result = preparedStatement.executeUpdate();
            logger.trace("saved {} instances", result);

            ResultSet resultSet = preparedStatement.getGeneratedKeys();
            if (resultSet.next()) {
                Long id = resultSet.getLong(1);
                entity.setId(id);
                logger.traceExit(entity);
                return entity;
            }

            return null;
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    @Override
    public Participant deleteById(Long entityId) {
        return null;
    }

    @Override
    public Participant update(Participant entity) {
        return null;
    }
}
